documentation/projects/proposals/rekognition_data/20240530-implementation_plan_augment_catalog_with_rekognition_tags.md

```diff
@@ -0,0 +1,669 @@
+        "Instances": [],
+        "Parents": []
+      },
+      {
+        "Name": "Party",
+        "Confidence": 57.07652282714844,
+        "Instances": [],
+        "Parents": []
+      },
+      {
+        "Name": "Dinner",
+        "Confidence": 56.07081985473633,
+        "Instances": [],
+        "Parents": [
+          {
+            "Name": "Food"
+          }
+        ]
+      },
+      {
+        "Name": "Supper",
+        "Confidence": 56.07081985473633,
+        "Instances": [],
+        "Parents": [
+          {
+            "Name": "Food"
+          }
+        ]
+      }
+    ],
+    "LabelModelVersion": "2.0",
+    "ResponseMetadata": {
+      "RequestId": "60c4b6f5-3b73-466e-8fa5-e40037661253",
+      "HTTPStatusCode": 200,
+      "HTTPHeaders": {
+        "content-type": "application/x-amz-json-1.1",
+        "date": "Thu, 29 Oct 2020 19:46:02 GMT",
+        "x-amzn-requestid": "60c4b6f5-3b73-466e-8fa5-e40037661253",
+        "content-length": "3526",
+        "connection": "keep-alive"
+      },
+      "RetryAttempts": 0
+    }
+  }
+}
+```
+
+This file is about 200GB in total. For more information about the data, see
+[Analysis Explanation](#analysis-explanation).
+
+#### DAG
+
+```{attention}
+A snapshot of the catalog database should be created prior to running this step
+in production.
+```
+
+We will create a DAG (`add_rekognition_labels`) which will perform the following
+steps:
+
+1. Create a temporary table in the catalog for storing the tag data. This table
+   will be two columns: `identifier` and `tags` (with data types matching the
+   existing catalog columns).
+2. Iterate over the large Rekognition dataset in a chunked manner using
+   [`smart_open`][smart_open]. `smart_open` provides
+   [options for tuning buffer size](https://github.com/piskvorky/smart_open?tab=readme-ov-file#transport-specific-options)
+   so larger chunks can be read into memory.
+   1. For each line, read in the JSON object and pull out the top-level labels &
+      confidence values. **Note**: some records may not have any labels.
+   2. Filter out any labels that are below the
+      [accuracy threshold](#accuracy-selection) and that don't match the
+      [excluded labels](#determine-excluded-labels).
+   3. Construct a `tags` JSON object similar to the existing tags data for that
+      image, including accuracy and provider. Ensure that the labels are lower
+      case and that the confidence value is between 0.0 and 1.0 (e.g.
+      `[{"name": "cat", "accuracy": 0.9983, "provider": "rekognition"}, ...]`).
+   4. At regular intervals, insert batches of constructed `identifier`/`tags`
+      pairs into the temporary table.
+3. Launch a [batched update run][batched_update] which merges the existing tags
+   and the new tags from the temporary table for each
+   identifier[^batch_tag_example]. **Note**: the batched update DAG may need to
+   be augmented in order to reference data from an existing table, similar to
+   #3415.
```

Comment:
FWIW I think the select/update queries could be written to reference an existing table, particularly if we use `resume_update`/`query_id` carefully -- but still a good thing to note, since we could definitely add support in a less workaround way if we anticipate keeping this DAG around for future tag updates :)